torch/export/dynamic_shapes.py

```diff
@@ -394,17 +569,86 @@ def tree_zip(combined_args, dynamic_shapes):
                     f"got {dynamic_shapes} instead",
                 )
 
+    from collections import defaultdict
+
+    # map of Dim names representing input shape dimensions to constraints on them
+    symbols: Dict[str, List[Constraint]] = defaultdict(list)
+    # track roots that do not directly represent input shape dimensions
+    phantom_roots: Dict[str, _PhantomRoot] = {}
+    derived_constraints_with_phantom_root: List[_DerivedConstraint] = []
+
     def to_constraint(dim, tensor, i):
-        constraint = dynamic_dim(tensor, i, debug_name=dim.__name__)
-        if dim.min != 2:
-            constraint = constraint >= dim.min
-        if dim.max != sys.maxsize - 1:
-            constraint = constraint <= dim.max
-        return constraint
+        import sympy
 
-    from collections import defaultdict
+        from torch.fx.experimental.symbolic_shapes import StrictMinMaxConstraint
+        from torch.utils._sympy.solve import try_solve
+        from torch.utils._sympy.value_ranges import ValueRanges
+
+        def root_value():
+            # given tensor.shape[i] is the value of dim = fn(root),
+            # find the value of root
+            symbol = sympy.Symbol(dim.root.__name__, integer=True)
```

Comment:
I'll clamp the bounds. We haven't yet moved the bounds to `>= 0` by the way, that's a TODO (somebody else is working on a PR for that).

Reply:
To be clear, I'm not going to use `nonnegative` (but instead enforce it via bounds) because we have some work TODO on serde to carry assumptions on all symbols in shape env, and I don't want to add assumptions that cannot be maintained through serde. (The only option is to also blanket-use `nonnegative` there right now, which is not correct for unbacked symints. I have a comment on that in this PR near serde logic.)